SQLServer XE handlers documentation (#29291)
DataDog/documentation: content/en/database_monitoring/guide/sql_extended_events.md

        duration > 1000000 -- in microseconds, limit to queries with duration greater than 1 second
    )
),
ADD EVENT sqlserver.module_end( -- capture stored procedure completions
    SET collect_statement = (1)
    ACTION ( -- datadog requires these exact actions for module_end
        sqlserver.sql_text,
        sqlserver.database_name,
        sqlserver.username,
        sqlserver.client_app_name,
        sqlserver.client_hostname,
        sqlserver.session_id,
        sqlserver.request_id
    )
    WHERE (
        sql_text <> '' AND
        duration > 1000000 -- in microseconds, limit to queries with duration greater than 1 second
    )
)
ADD TARGET package0.ring_buffer -- do not change, datadog is only configured to read from ring buffer at this time
WITH (
    MAX_MEMORY = 2048 KB, -- do not exceed 2048, values above 2 MB may result in data loss due to SQLServer internals
    TRACK_CAUSALITY = ON, -- allows datadog to correlate related events across activity ID
    EVENT_RETENTION_MODE = ALLOW_SINGLE_EVENT_LOSS,
    MAX_DISPATCH_LATENCY = 30 SECONDS,
    STARTUP_STATE = ON
);

ALTER EVENT SESSION datadog_query_completions ON SERVER STATE = START;
GO
```

> The datadog_query_errors XE session captures SQL errors (severity ≥ 11) and query timeouts (attention events), allowing Datadog to surface query failures and interruptions.

Reviewer: Are there any docs for severity >= 11?

Reviewer: Also what are attention events?

Author: Adding this link: https://learn.microsoft.com/en-us/sql/relational-databases/errors-events/database-engine-error-severities
And this link: https://learn.microsoft.com/en-us/sql/relational-databases/event-classes/attention-event-class

ADD EVENT sqlserver.error_reported(
    ACTION( -- datadog requires these exact actions for error_reported
        sqlserver.sql_text,
        sqlserver.database_name,
        sqlserver.username,
        sqlserver.client_app_name,
        sqlserver.client_hostname,
        sqlserver.session_id,
        sqlserver.request_id
    )
    WHERE severity >= 11
),
ADD EVENT sqlserver.attention(
    ACTION( -- datadog requires these exact actions for attention
        sqlserver.sql_text,
        sqlserver.database_name,
        sqlserver.username,
        sqlserver.client_app_name,
        sqlserver.client_hostname,
        sqlserver.session_id,
        sqlserver.request_id
    )
)
ADD TARGET package0.ring_buffer -- do not change, datadog is only configured to read from ring buffer at this time
WITH (
    MAX_MEMORY = 2048 KB, -- do not change, setting this larger than 2 MB may result in data loss due to SQLServer internals
    EVENT_RETENTION_MODE = ALLOW_SINGLE_EVENT_LOSS,
    MAX_DISPATCH_LATENCY = 30 SECONDS,
    STARTUP_STATE = ON
);

ALTER EVENT SESSION datadog_query_errors ON DATABASE STATE = START;
GO
```
{{% /tab %}}

{{< /tabs >}}

> 2. In the Datadog Agent, enable xe_collection in `sqlserver.d/conf.yaml`.

Reviewer 2: Can we add a link to the [conf.yaml.example](https://github.com/DataDog/integrations-core/blob/master/sqlserver/datadog_checks/sqlserver/data/conf.yaml.example) file here (or somewhere on this page)? As this is a separate guide, it would help users avoid jumping back and forth to find it in the DBM setup guide

Reviewer: ^ agreed

```yaml
  xe_collection:
    query_completions:
      enabled: true
    query_errors:
      enabled: true
```

> In order to collect deobfuscated versions of query_completion and query_error events, enable collect_raw_query_statement in `sqlserver.d/conf.yaml`.

Reviewer: We should probably add this disclaimer that we include above the param in sample conf file here the docs as well: https://github.com/DataDog/integrations-core/blob/master/sqlserver/datadog_checks/sqlserver/data/conf.yaml.example#L659

"  ## Raw query statements and execution plans may contain sensitive information (e.g., passwords)
    ## or personally identifiable information in query text.
    ## Enabling this option will allow the collection and ingestion of raw query statements and
    ## execution plans into Datadog, which can then become viewable in query samples or explain plans.
    ## This option is disabled by default."

    EVENT_RETENTION_MODE = ALLOW_SINGLE_EVENT_LOSS,
    MAX_DISPATCH_LATENCY = 30 SECONDS,
    STARTUP_STATE = ON
);

ALTER EVENT SESSION datadog_query_errors ON SERVER STATE = START;
GO
```

2. In the Datadog Agent, enable xe_collection in `sqlserver.d/conf.yaml`.
See the [sample conf.yaml.example][2] for all available configuration options.
```yaml
  xe_collection:
    query_completions:
      enabled: true
    query_errors:
      enabled: true
```
In order to collect deobfuscated versions of query_completion and query_error events, enable collect_raw_query_statement in `sqlserver.d/conf.yaml`. Please note, raw query statements and execution plans may contain sensitive information (e.g., passwords) or personally identifiable information in query text. Enabling this option will allow the collection and ingestion of raw query statements and execution plans into Datadog, which can then become viewable in query samples or explain plans. This option is disabled by default.
```yaml
  collect_raw_query_statement:
    enabled: true
```

{{% /tab %}}

{{% tab "Azure DB" %}}

> 1. In the SQL Server database, create the following Extended Events (XE) sessions:

Reviewer 3: ```suggestion
1. In your Azure SQL Server Database, create the following Extended Events (XE) sessions:
```

> The datadog_query_completions XE session captures long-running SQL queries (over 1 second) from RPC calls, SQL batches, and stored procedures.

Reviewer 3: ```suggestion
The `datadog_query_completions` XE session captures long-running SQL queries (over 1 second) from RPC calls, SQL batches, and stored procedures.
```

```sql

> -- 1. Query completions (grouped)
> -- Includes RPC completions, batch completions, and stored procedure completions

Reviewer 3: ```suggestion
-- Query completions: RPC, batch, and stored procedure events
```

        duration > 1000000 -- in microseconds, limit to queries with duration greater than 1 second
    )
),
ADD EVENT sqlserver.module_end( -- capture stored procedure completions
    SET collect_statement = (1)
    ACTION ( -- datadog requires these exact actions for module_end
        sqlserver.sql_text,
        sqlserver.database_name,
        sqlserver.username,
        sqlserver.client_app_name,
        sqlserver.client_hostname,
        sqlserver.session_id,
        sqlserver.request_id
    )
    WHERE (
        sql_text <> '' AND
        duration > 1000000 -- in microseconds, limit to queries with duration greater than 1 second
    )
)
ADD TARGET package0.ring_buffer -- do not change, datadog is only configured to read from ring buffer at this time
WITH (
    MAX_MEMORY = 2048 KB, -- do not exceed 2048, values above 2 MB may result in data loss due to SQLServer internals
    TRACK_CAUSALITY = ON, -- allows datadog to correlate related events across activity ID
    EVENT_RETENTION_MODE = ALLOW_SINGLE_EVENT_LOSS,
    MAX_DISPATCH_LATENCY = 30 SECONDS,
    STARTUP_STATE = ON
);

ALTER EVENT SESSION datadog_query_completions ON DATABASE STATE = START;
GO
```

> The datadog_query_errors XE session captures SQL errors of [severity ≥ 11](https://learn.microsoft.com/en-us/sql/relational-databases/errors-events/database-engine-error-severities) and query timeouts, called [attention events](https://learn.microsoft.com/en-us/sql/relational-databases/event-classes/attention-event-class), allowing Datadog to surface query failures and interruptions.

Reviewer 3: ```suggestion
The datadog_query_errors XE session captures SQL errors of [severity ≥ 11][1] and query timeouts (also known as [attention events][2]), enabling Datadog to report query failures and timeouts.
```

        sqlserver.request_id
    )
    WHERE (
        sql_text <> '' AND
        duration > 1000000 -- in microseconds, limit to queries with duration greater than 1 second
    )
),
ADD EVENT sqlserver.module_end( -- capture stored procedure completions
    SET collect_statement = (1)
    ACTION ( -- datadog requires these exact actions for module_end
        sqlserver.sql_text,
        sqlserver.database_name,
        sqlserver.username,
        sqlserver.client_app_name,
        sqlserver.client_hostname,
        sqlserver.session_id,
        sqlserver.request_id
    )
    WHERE (
        sql_text <> '' AND
        duration > 1000000 -- in microseconds, limit to queries with duration greater than 1 second
    )
)
ADD TARGET package0.ring_buffer -- do not change, datadog is only configured to read from ring buffer at this time
WITH (
    MAX_MEMORY = 2048 KB, -- do not exceed 2048, values above 2 MB may result in data loss due to SQLServer internals
    TRACK_CAUSALITY = ON, -- allows datadog to correlate related events across activity ID
    EVENT_RETENTION_MODE = ALLOW_SINGLE_EVENT_LOSS,
    MAX_DISPATCH_LATENCY = 30 SECONDS,
    STARTUP_STATE = ON
);
GO

> -- 2. Errors and Attentions (grouped)

Reviewer 4: Would it make sense to split the 2 sessions into 2 separate code blocks? Or at least mentioning the 2 different session types outside the code block (since they are mentioned on line 190)

Author: I've split into 2 sections

  - Fewer queries: Increase the threshold (e.g., `duration > 5000000` for 5 seconds)
- **Important**: Setting thresholds too low can result in:
  - Excessive event collection affecting server performance
  - Event loss due to buffer overflow
  - Incomplete data as Datadog only collects the most recent 1000 events per collection interval

**Memory Allocation**
- Default is `MAX_MEMORY = 2048 KB`.
- Don't exceed 2048 KB as larger values may cause data loss due to SQL Server internal limitations
- For high-volume servers, consider keeping this at maximum (2048 KB)
- For lower traffic servers, 1024 KB may be sufficient

**Event Filtering**
- Add additional filters to the WHERE clause to reduce event volume:
  ```sql
  WHERE (
      sql_text <> '' AND
      duration > 1000000 AND
      -- Add custom filters here
      database_name = 'YourImportantDB' -- Only track specific databases
      -- OR --
      username <> 'ReportUser' -- Exclude specific users
  )
  ```

**Performance Considerations**
- Extended Events are designed to be lightweight but still have some overhead
- If you observe performance issues, consider:

>   - Increasing duration thresholds

Reviewer 4: I think linking back up the page to how to do this would be good

>   - Adding more specific filters

Reviewer 4: similar to my previous comment

>   - Disabling one or both sessions during peak load periods

Reviewer 4: naming the sessions or being able to explicit how to would be great IMO

        sqlserver.request_id
    )
    WHERE (
        sql_text <> '' AND
        duration > 1000000 -- in microseconds, limit to queries with duration greater than 1 second
    )
),
ADD EVENT sqlserver.module_end( -- capture stored procedure completions
    SET collect_statement = (1)
    ACTION ( -- datadog requires these exact actions for module_end
        sqlserver.sql_text,
        sqlserver.database_name,
        sqlserver.username,
        sqlserver.client_app_name,
        sqlserver.client_hostname,
        sqlserver.session_id,
        sqlserver.request_id
    )
    WHERE (
        sql_text <> '' AND
        duration > 1000000 -- in microseconds, limit to queries with duration greater than 1 second
    )
)
ADD TARGET package0.ring_buffer -- do not change, datadog is only configured to read from ring buffer at this time
WITH (
    MAX_MEMORY = 2048 KB, -- do not exceed 2048, values above 2 MB may result in data loss due to SQLServer internals
    TRACK_CAUSALITY = ON, -- allows datadog to correlate related events across activity ID
    EVENT_RETENTION_MODE = ALLOW_SINGLE_EVENT_LOSS,
    MAX_DISPATCH_LATENCY = 30 SECONDS,
    STARTUP_STATE = ON
);
GO

> -- 2. Errors and Attentions (grouped)

Reviewer 4: Same comment as for the default SQL Server (if splitting makes sense)

  ```sql
  WHERE (
      sql_text <> '' AND
      duration > 1000000 AND
      -- Add custom filters here
      database_name = 'YourImportantDB' -- Only track specific databases
      -- OR --
      username <> 'ReportUser' -- Exclude specific users
  )
  ```

**Azure-Specific Considerations**
- In Azure SQL Database, resources are more constrained
- Consider using more restrictive filtering if you're on lower-tier service levels
- For elastic pools, monitor performance impact across all databases

2. In the Datadog Agent, enable xe_collection in `sqlserver.d/conf.yaml`.
```yaml
  xe_collection:
    query_completions:
      enabled: true
    query_errors:
      enabled: true
```
In order to collect deobfuscated versions of query_completion and query_error events, enable collect_raw_query_statement in `sqlserver.d/conf.yaml`.
```yaml
  collect_raw_query_statement:
    enabled: true
```

{{% /tab %}}

{{< /tabs >}}

## Exploring query errors

To access the query error view, navigate to the **APM** > **Database Monitoring** > **Samples** tab, and use the SQLServer In Query Errors dropdown selector to view a list of Query Errors

## Exploring deobfuscated queries

> Workflow is still a work in progress.

Reviewer 4: Is this expected to be released with this? Or will the docs be updated to reflect the actual workflow? Because if the workflow is indeed a WIP by the time the docs are released, I think we need a different way to phrase this
